### PR TITLE
[WIP] Run test daemons in a network namespace.

### PR DESCRIPTION
**Pull request by cpuguy83**
This isolates test daemons further, particularly for global resources
such as iptables and port-space.

### CODE REVIEW

File: internal/test/daemon/daemon_unix.go [hunk 1]
```diff
@@ -48,3 +54,105 @@ func SignalDaemonDump(pid int) {
...
 }
+
+var daemonNetworks cni.CNI
+var daemonNetworksOnce sync.Once
+
+func ensureDaemonNetworks() error {
+	var err error
+	daemonNetworksOnce.Do(func() {
+		daemonNetworks, err = cni.New(cni.WithLoNetwork, cni.WithDefaultConf)
+		errors.Wrap(err, "error setting up cni for daemon networks")
```
Comment (AkihiroSuda) [inline internal/test/daemon/daemon_unix.go]: `err = errors.Wrap(err, ...)`?

Comment (cpuguy83) [reply internal/test/daemon/daemon_unix.go]: Nice catch, fixed :)

**Comment by cpuguy83** [comment]
Ok, generally pretty close here *but* I am running into issues with things hitting localhost.
It seems I'm in the netns but anything with localhost (a daemon, a registry, whatever) don't seem to work so something's not right there.
Before I spend too much more time debugging this (or maybe someone else can point out what's going on), PTAL and let me know if this is worth the change.

ping @tonistiigi @thaJeztah @arkodg @tiborvass

**Comment by tonistiigi** [comment]
@cpuguy83 Something equivalent to https://github.com/moby/moby/pull/24613/files#diff-166c8d9926e36ed276ab9ccd7c150809R610 https://github.com/moby/moby/pull/24613/files#diff-a4a600e2f68a7483ea3b27fa19db43eeR31 ?

**Comment by cpuguy83** [comment]
@tonistiigi There are a couple of places where I have a proxy server setup. Some TLS tests make this a bit annoying.

**Comment by cpuguy83** [comment]
Is it expected that janky is not running integration tests, @thaJeztah @tiborvass

**Comment by cpuguy83** [comment]
Nevermind I forgot I cherry-picked in a fix for the flaky test detector and that fix was apparently not right.

**Comment by cpuguy83** [comment]
Well I think this would be mostly ready except tests are taking WAY longer to run and I do not know why.
The namespace configuration overhead is negligible (for both creating and deleting), and the actual daemon startup (process start -> API ready) is also minimal.
Tests are communicating over the unix socket so not even bridge overhead there.

The slowest netns config I've seen is 4s (that is creating the ns and setting up the bridge) and that is an outlier, most are under 1s, meanwhile tests that should take 2s are taking 12 (as an example).
This is extremely evident in the integration/service tests (which are timing out).

Example from this PR:
```
--- PASS: TestDockerNetworkReConnect (21.41s)
    network_test.go:83: Creating a new daemon
    net_linux.go:62: [d3541f40c37fb] creating network namespace
    net_linux.go:75: [d3541f40c37fb] setting up cni networking
    net_linux.go:49: [d3541f40c37fb] configureNetNS duration: 2.269827
    daemon.go:363: [d3541f40c37fb] waiting for daemon to start
    daemon.go:363: [d3541f40c37fb] waiting for daemon to start
    daemon.go:391: [d3541f40c37fb] daemon started
    daemon.go:324: [d3541f40c37fb] daemon startup time: 0.740407
    daemon.go:501: [d3541f40c37fb] Stopping daemon
    daemon.go:331: [d3541f40c37fb] exiting daemon
    daemon.go:488: [d3541f40c37fb] Daemon stopped
    net_linux.go:98: [d3541f40c37fb] cleaning up cni networking
    net_linux.go:103: [d3541f40c37fb] removing network namespace
    net_linux.go:96: [d3541f40c37fb] netns cleanup duration: 0.182375s
```

And another PR:
```
11:56:06 --- PASS: TestDockerNetworkReConnect (3.31s)
11:56:06     network_test.go:82: Creating a new daemon
11:56:06     daemon.go:336: [d2859c14b39fa] waiting for daemon to start
11:56:06     daemon.go:336: [d2859c14b39fa] waiting for daemon to start
11:56:06     daemon.go:364: [d2859c14b39fa] daemon started
11:56:06     daemon.go:472: [d2859c14b39fa] Stopping daemon
11:56:06     daemon.go:307: [d2859c14b39fa] exiting daemon
11:56:06     daemon.go:459: [d2859c14b39fa] Daemon stopped
```

**Comment by cpuguy83** [comment]
Seems to be that `Cleanup` calls `Stop` now and we always call cleanup.

**Comment by cpuguy83** [comment]
🙁 

```
16:36:06 --- PASS: TestServiceUpdateConfigs (44.53s)
16:36:06     update_test.go:141: Creating a new daemon
16:36:06     net_linux.go:60: [df6f03afbbf1a] creating network namespace
16:36:06     net_linux.go:73: [df6f03afbbf1a] setting up cni networking
16:36:06     daemon.go:262: [df6f03afbbf1a] Daemon configureNetNS duration: 0.039535s
16:36:06     daemon.go:262: [df6f03afbbf1a] Daemon configureNetNS duration: 0.000000s
16:36:06     daemon.go:375: [df6f03afbbf1a] waiting for daemon to start
16:36:06     daemon.go:375: [df6f03afbbf1a] waiting for daemon to start
16:36:06     daemon.go:403: [df6f03afbbf1a] daemon started
16:36:06     daemon.go:336: [df6f03afbbf1a] daemon startup time: 0.507045
16:36:06     daemon.go:262: [df6f03afbbf1a] Daemon daemon start duration: 0.507097s
16:36:06     daemon.go:514: [df6f03afbbf1a] Stopping daemon
16:36:06     daemon.go:343: [df6f03afbbf1a] exiting daemon
16:36:06     daemon.go:501: [df6f03afbbf1a] Daemon stopped
16:36:06     daemon.go:262: [df6f03afbbf1a] Daemon daemon stop duration: 0.481983s
16:36:06     net_linux.go:93: [df6f03afbbf1a] cleaning up cni networking
16:36:06     net_linux.go:98: [df6f03afbbf1a] removing network namespace
16:36:06     daemon.go:262: [df6f03afbbf1a] Daemon daemon nsnet cleanup duration: 0.053191s
16:36:06     daemon.go:262: [df6f03afbbf1a] Daemon cleanup duration: 0.535373s
```

As compared to master:

```
11:57:09 --- PASS: TestServiceUpdateConfigs (14.67s)
11:57:09     update_test.go:139: Creating a new daemon
11:57:09     daemon.go:336: [def447ee5e7b6] waiting for daemon to start
11:57:09     daemon.go:336: [def447ee5e7b6] waiting for daemon to start
11:57:09     daemon.go:364: [def447ee5e7b6] daemon started
11:57:09     daemon.go:472: [def447ee5e7b6] Stopping daemon
11:57:09     daemon.go:307: [def447ee5e7b6] exiting daemon
11:57:09     daemon.go:459: [def447ee5e7b6] Daemon stopped
```

**Comment by cpuguy83** [comment]
Going to close this as there's some as of yet unexplained performance impact here, it also ended up being a really complex change...

Maybe someone else wants to pick this up?

**Pull request closed without merging**